Add host tool

mzgoddard opened the pull request:
- Add `src/agent/README.md` with some "how to use" information on agent
- Add `bin/host.js` and `src/host` directory
- Add `src/host/README.md` with some "how to use"  information on host
- Add `src/host/cli-main.js` defining the host's main command as a yargs module
- Add `src/host/cli-plan.js` defining the host's plan command as a yargs module
- `src/host/agent.js` starts and manages an agent instance through child_process or js api
- `src/host/plan-from.js` loads a test plan from a collection of files through child_process or js api

jugglinmike commented:
@z I've begun to review this as requested, but I see that you're still pushing up changes. Does this need more time before Simon or I take a look?

mzgoddard commented:
> @z I've begun to review this as requested, but I see that you're still pushing up changes. Does this need more time before Simon or I take a look?

The changes I've made today are related to enabling and fixing CI on linux and windows. I'm limiting any changes I make at the moment to small tweaks like that to assist in the review process. If tests themselves are failing, I'll leave fixing that until you'll can take a look. So please take a look when you can.

I want to add more tests, but until it makes sense, I'll do that on another branch.

jugglinmike commented on src/host/README.md:
````diff
@@ -0,0 +1,122 @@
+A short invocation of the main command can be made with a list of files or file globs that make up a test plan. A test plan should contain tests for a single assistive technology to be tested. If for example all the reference files and test files are in two directories this command could be executed.
+
+```sh
+$ bin/host.js reference/** at/** >result.json
````
I've been typing things like `./bin/host.js` all my life because I never knew that the presence of a forward slash influences how shells resolve the command. Thanks for the tip!

jugglinmike reviewed (commented):
Hey @mzgoddard! A few high-level things:

As we discussed last week, since the "protocol" abstraction is not motivated by a specific use case, and since implementing it involves so much extra code, it would be better to remove it for now.

Some files include a comment that's consistent with [TypeScript's "triple-slash directives" feature](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html), though there are no TypeScript-related tools in the project's development dependencies. Are there any additional tools that other contributors should use to participate in this project?

My appreciation for the UNIX philosophy notwithstanding, I'm wondering if all the advanced usages (i.e. those involving input/output redirection) are feasible on Windows Command Prompt and Powershell. Part of the issue is that the proposed documentation assumes a POSIX environment, but the more fundamental part is whether this interface is even usable on the other platforms. Do you know if folks using those other shells can accomplish the same tasks? Or do you think we'll need to extend the CLI with dedicated arguments?

mzgoddard commented on src/host/server.js:
```diff
@@ -0,0 +1,124 @@
+    this._directories = new Map();
+
+    this._app = express();
+    this._app.use('/:directory', this._request.bind(this));
```
`use` and `get` do not have the same behavior here. Namely in this case how they manipulate url members of the request object. I used `use` here because it will remove the matching part of the url.

mzgoddard commented on src/agent/README.md:
````diff
@@ -0,0 +1,57 @@
+The agent tool can be invoked to perform tests from json files by piping them into agent. The result will be emitted out to the command's output stream.
+
+```sh
+$ cat at/test-1-jaws.collected.json | bin/agent.js >result.json
````
I added `src/data/types.js` to locate such types there.

Using TAP would be really interesting. I'd like to take a shot at that some time.

mzgoddard commented on src/agent/mock-test-runner.js:
```diff
@@ -1,38 +1,97 @@
-        task.assertions.map(assertion => this.runAssertion(command, assertion))
+      results: await task.commands.reduce(
+        async (carry, command) => [
+          ...(await carry),
```
I'm not sure why you are suggesting removing `await`. `carry` is a promise. I've changed the initial value to be a promise resolving to an empty array.

mzgoddard commented on src/host/README.md:
```diff
@@ -0,0 +1,122 @@
+
+This directory implements the systems used by the `bin/host.js` command line tool.
+
+The `bin/host.js` tool has two commands. The main command operates a server and `bin/agent.js` instance to run test plans and the `plan` subcommand reads files from disk and packages them for the main command.
```
I named the main command `run-plan` and renamed `plan` to `read-plan`.

mzgoddard commented on src/host/README.md:
```diff
@@ -0,0 +1,122 @@
+$ plans=(
+array> breadcrumb
+array> checkbox
+array> )
```
In basic bash it would render as

```sh
$ plans=(
> breadcrumb
> checkbox
> )
```

I had included the shell prompts which in `()` lists starts with `>` normally. My shell configuration had changed that `array>`.

mzgoddard commented on src/host/agent.js:
```diff
@@ -0,0 +1,295 @@
+/// <reference path="types.js" />
+
+import child_process from 'child_process';
+import { SIGINT } from 'constants';
```
Replaced with

```
import {constants as osConstants} from 'os';
const {signals: {SIGINT}} = osConstants;
```

mzgoddard commented on src/host/agent.js:
```diff
@@ -0,0 +1,295 @@
+  fork: [AgentFork],
+  shell: [AgentShell],
+  api: [AgentAPI],
+  auto: [AgentFork, AgentShell, AgentAPI],
```
Hmm. From host's perspective, fork and shell have the same use case. API has a slightly different use case, trading the ability to recover from agent crashing for performance. Granted `AgentController` doesn't have recovery implemented yet.

mzgoddard replied:
`auto`'s use case is kind of a small extension, maybe it should be called an experiment, in recovery. When the agent is started if there is an issue, `auto` will try the next in line, until one successfully starts. Previously `auto` was the default, but I changed it to `fork` since including `api` in it would limit the ability to recover

mzgoddard commented on src/host/agent.js:
```diff
@@ -0,0 +1,295 @@
+      config: { debug: config.quiet || config.verbose ? false : true, ...config },
+    };
+    this._instanceOptions = null;
+    this._instance = null;
```
I renamed `Agent` to `AgentController` and created a parent for `AgentFork` and `AgentAPI` called `AgentProtocol`. Maybe `AgentProcess` would be a better name than `AgentProtocol`.

`AgentController` is the point of recovery if agent crashes. As a separation of concerns `AgentProtocol` implementations deal with communication to `src/agent` code and `AgentController` tries to recover from `Agent` crashes.

mzgoddard replied:
I tried thinking about how to rename `_instance`, I'm not sure what to do for that yet.

mzgoddard commented on src/host/main.js:
```diff
@@ -0,0 +1,72 @@
+      });
+
+      const file = plan.files.find(({ name }) => name === test.filepath);
+      const result = await agent.run(JSON.parse(Buffer.from(file.buffer).toString()));
```
I don't think `buffer` is a misleading name. `buffer` here refers to it being a portion of memory.

I updated the documentation references to state `TestPlan#files` as `FileRecord.NamedRecord[]`.

mzgoddard commented on src/shared/messages.js:
```diff
@@ -11,7 +11,7 @@ import { EventEmitter } from 'events';
  * @template T
  */
-export function createLogger(messages) {
+export function _createLogger(messages) {
```
Renamed this to `createSharedLogger`. I was updating names of some things to avoid `import {thing as prefixedThing}` cases. I chose `_createLogger` I think to in part indicate that its use is restricted to `aria-at-harness`, and rename it for when `src/agent/messages.js` and `src/host/messages.js` import it without having to rename it.

mzgoddard commented on src/host/plan-object.js:
```diff
@@ -0,0 +1,176 @@
+ * @param {*} url
+ * @returns {AriaATCIShared.BaseURL}
+ */
+function validateTestPlanURL(url) {
```
I wish I could I use `typescript` here. I've tried to not venture away from how `aria-at` operates in regards to dependencies. Such as minimal use of external dependencies and build steps. The returns here are editor integration workaround. I have typescript integration enabled in my editor of choice, but using jsdoc limits that, in this spot that jsdoc doesn't have a `url asserts BaseURl` like return type.

mzgoddard commented on src/shared/file-record.js:
```diff
@@ -0,0 +1,516 @@
+   * @param {string} glob a fs-like glob to test paths with
+   * @returns {function(string): boolean}
+   */
+  function compileGlob(glob) {
```
This is part, limit-external-dependencies mindset and most libraries with glob in name or keywords provide a read-files-based-on-glob api and don't provide their matching algorithm. I know looking previously I saw minimatch but it hasn't seen a patch in a few years so I guess I'd need to look at that too.

mzgoddard commented on src/host/plan-from.js:
```diff
@@ -0,0 +1,177 @@
+planFromCommandShell.protocol = 'shell';
+
+async function planFromAPI({ workingdir, files }) {
+  const host = createHost({ path: path.posix });
```
Ah. Yes this was an error on my part. Passing path.posix was meant to be in just multipart so paths in multipart were independent of os. I incorrectly used that concept here.

mzgoddard commented on src/shared/file-record.js:
```diff
@@ -0,0 +1,516 @@
+  return Boolean(record && 'buffer' in record && record.buffer);
+}
+
+function createHost({ fs = _fs.promises, path = _path } = {}) {
```
I think I understand what you are getting at but I'm not sure how this isn't OOP.

I think I opted for this style because the records are data structures with no methods. As well the built host doesn't have internal state between the set of exposed functions. That `fs` and `path` are similar basic object collections of functions probably weighed into it as well.

mzgoddard commented on src/shared/file-record.js:
```diff
@@ -0,0 +1,516 @@
+   * @param {FileRecord.Record} next
+   * @returns {FileRecord.Change}
+   */
+  function changesFrom(previous, next) {
```
Yup. Removed it and other similar king and `file-record-chain.js`.

mzgoddard commented on src/shared/job.js:
```diff
@@ -0,0 +1,74 @@
+              if (iterator.throw) {
+                return await iterator.throw(...args);
+              }
+              return { done: true };
```
Yes.

mzgoddard commented on src/shared/iterate-emitter.test.js:
```diff
@@ -0,0 +1,141 @@
+  );
+});
+
+test('iterate queued values emitted by an EventEmitter after it emits a completion event', async t => {
```
`iterate queued ...` is concerned with the values before `emit('complete')` and `stops iterating ...` is concerned with the values after `emit('complete')`. `stop iterating ...` has a timeout after each emitted value such that when it `complete`s iterateEmitter's value queue will have been consumed. `iterate queued ...` without a timeout will `complete` when it's queue has 5 values.

Additionally `stop iterating ...` does not have a timeout after `complete`. That's important so that `emit('next', 5)` occurs. That will run iterateEmitter's next listener which should not queue 5 since `complete` has happened.

It is true that `emit('next', 6)` shouldn't happen. Maybe that second loop should be changed. But I'm not sure.

jugglinmike reviewed (changes requested):
Looking good, @mzgoddard! I'm moving a few topics out here to the main discussion thread since it's a little hard to find the corresponding inline comments and they're not related to a single line of code.

> > jugglinmike 24 days ago
> >
> > Do these protocols each have a different use case? Also, under what circumstances would we want to use "auto"?
>
> mzgoddard 2 hours ago
>
> Hmm. From host's perspective, fork and shell have the same use case. API has a slightly different use case, trading the ability to recover from agent crashing for performance. Granted `AgentController` doesn't have recovery implemented yet.
>
> `auto`'s use case is kind of a small extension, maybe it should be called an experiment, in recovery. When the agent is started if there is an issue, `auto` will try the next in line, until one successfully starts. Previously `auto` was the default, but I changed it to `fork` since including `api` in it would limit the ability to recover

I asked about `auto` because for a controlled testing environment like we're building, I don't know if implicit recovery mechanisms are desirable. Until we have real-world experience running the system, I'm considering any protocol-level failure to be an exceptional circumstance--one that maintainers should be informed about immediately. Although a fallback is ostensibly more robust, it will make it harder to recognize when something failed (leading to situations where we might say, "this hasn't worked for three months!") and harder to reason about what's actually going on in continuous integration (causing uncertainty like, "this new error might be related to XYZ, but only if the 'API' protocol was in use.").

This is less of a concern now that `auto` is no longer the default, but experiments are generally best conducted on feature branches. That reduces the "sharp edges" exposed to end-users, and it limits the complexity of the code we're actually shipping.

Separately, per [my earlier question](https://github.com/bocoup/aria-at-harness/pull/2#pullrequestreview-812892655), "Are there any additional tools that other contributors should use to participate in this project?"

> I have typescript integration enabled in my editor of choice

At the moment, we're not giving contributors any guidance about how they can enjoy the same advantages from the available type annotations. (I'm not sure I could personally replicate your setup, either.) I think that if we feel those annotations are useful enough to belong in the source code, then it's our responsibility to show others how to use them. One risk of leaving this undocumented is that we discourage contributors. Another risk is that future contributions unintentionally interfere with the tools that we're relying on.

jugglinmike reviewed (changes requested):
Looks like a wire got crossed somewhere; please see [my prior review](https://github.com/bocoup/aria-at-harness/pull/2#pullrequestreview-839561900) for my latest feedback.

mzgoddard commented on src/host/agent.js:
```diff
@@ -0,0 +1,336 @@
+        const ready = this._instance.start(options);
+        this._logEmitter.emit('log', this._instance);
+        await ready;
+        log(HostMessage.AGENT_PROTOCOL, { protocol: Protocol.protocol });
```
Changed it to `protocolName`. I tried `name` originally but `name` is already taken. Its the constructor/class name. So I figured should avoid that conflict.

mzgoddard commented on src/host/plan-from.js:
```diff
@@ -0,0 +1,124 @@
+  } else if (
+    typeof buffer === 'object' &&
+    buffer !== null &&
+    Object.keys(buffer).every(key => Number.isInteger(key))
```
This case was in use. I looked around to remember why. The serialization for the process was the default to `json,` I think I didn't use `advanced` because that was added in a v12 patch `v12.16.0`. I remove this function and switched to the advanced serialization. I can revert this change if we find we need to support `json` process serialization for some reason.

mzgoddard commented on src/host/plan-object.js:
```diff
@@ -0,0 +1,177 @@
+function validateUint8Array(buffer) {
+  invariant(
+    buffer instanceof Uint8Array,
+    () => `'${buffer && buffer.constructor.name}' === 'Uint8Array'`
```
> The difference between those two definitions is most apparent in code like this (which reads, "transform the buffer into a buffer"), and I think it may make the code harder to follow elsewhere.

I replaced the `Buffer.from(...).toString()` with a `TextDecoder.decode`.

> A name like bufferData would help distinguish between the Uint8Array which stores a portion of memory and the Buffer instance.

mzgoddard commented:
I made a branch with tests, [bocoup/aria-at-automation-harness#bocoup/test-host-tool](https://github.com/bocoup/aria-at-automation-harness/tree/ce02e5991543e25dfb832efb8457ff27c1d74eab), for this PR. I wasn't certain I should pushed them to this PR and complicate its review with them.

-----

> Until we have real-world experience running the system, I'm considering any protocol-level failure to be an exceptional circumstance--one that maintainers should be informed about immediately.

My main concern for being able to recover is for things arising from new or changed test plans. I expect that the harness will run on aria-at pull requests at some point and so may be fed tests that have some issue, and the automation harness will help identify that. That issue may be a typo from a test author or it could be a driver api and some error on our part we hadn't considered yet. I don't think this at that point yet, but I think that is where its going.

> Although a fallback is ostensibly more robust, it will make it harder to recognize when something failed (leading to situations where we might say, "this hasn't worked for three months!") and harder to reason about what's actually going on in continuous integration (causing uncertainty like, "this new error might be related to XYZ, but only if the 'API' protocol was in use.").

I think we would solve that by logging changes in state. Decisions the software makes like that should be available in logs. The protocol used for reading a test plan and controlling an agent are currently reported in logs. Those logs are not included in the default verbose level currently, but that is what the `--debug` flag is for. (I think we should make `--debug` default if we detect we are in a CI environment.)

I think that fallbacks would make "this hasn't worked for three months!" less likely. One mode could encounter a failure and another doesn't. Ultimately the software finishes and writes out what happened. I figure the `results` members in the `TestResult` types in `src/data` are arrays to include errors that were recovered from.

> This is less of a concern now that auto is no longer the default, but experiments are generally best conducted on feature branches. That reduces the "sharp edges" exposed to end-users, and it limits the complexity of the code we're actually shipping.

That's fair.

> Separately, per my earlier question, "Are there any additional tools that other contributors should use to participate in this project?"

>> I have typescript integration enabled in my editor of choice

> At the moment, we're not giving contributors any guidance about how they can enjoy the same advantages from the available type annotations. (I'm not sure I could personally replicate your setup, either.)

My editor is vscode with the builtin javascript/typescript functionality. Any tweaks I've made could be replicated by other users by including a `.vscode` directory in the repository that would contain those settings. But that runs up against the dependencies difficulty, which I'm not sure how to resolve without reaching some quorum with more people. (So I'll do just that with an Issue soon.)

> I think that if we feel those annotations are useful enough to belong in the source code, then it's our responsibility to show others how to use them.

I didn't consider that. I want to do this once we resolve dependencies.

> One risk of leaving this undocumented is that we discourage contributors.

Fair.

> Another risk is that future contributions unintentionally interfere with the tools that we're relying on.

I was thinking about that. I know what tools I would add to enforce this in PRs by failing CI. But ... dependencies.

-----

While writing this I couldn't respond directly to some of your comments. The reply field is missing for me. But only some of them. So I'll respond here.

-----

> It's now an instance of a Protocol, right? How about _protocol?

Erm. Something happened and I somehow ended up editing your comment instead of responding.

What do you think about `_activeProtocol`?

-----

> A limitation of your editor isn't a very convincing rationale for any software design decision :) That's especially true considering that other contributors will not have the same restrictions (though I have more to say about that beyond the context of this particular function).

Its a limitation of typescript's understanding of jsdoc and a limitation of jsdoc type system. I want to find or make an extension to provide a language server for editors for jsdoc but I don't that exists or could be made in a reasonable amount of time.

If someone didn't follow that limitation in contributions I would accept that and live with it unless we can depend on tools to enforce that in CI. I wouldn't require them to make style changes that aren't enforced by CI.

-----

> I'm not opposed to limiting dependencies (even though it means re-inventing solutions), but I also don't feel comfortable moving forward with an untested implementation of something that is both complex and straightforward to test.

I split the glob code into its own file. In my [test branch](https://github.com/bocoup/aria-at-automation-harness/blob/ce02e5991543e25dfb832efb8457ff27c1d74eab/src/shared/tests/file-glob.js), I added a lot of tests on it.

> (And for what it's worth, I'm starting to think that "time since last commit" is becoming a weaker signal as the Node.js module ecosystem matures. It seems feasible for libraries like minimatch to simply be complete.)

I agree. Its a useful metric because it can be quickly considered, but it doesn't really describe a project well, and so pretty limited. Same for semver.

-----

> By my reading, the host does have internal state: the function closure (which persists the values of fs and path). Written with the OOP style used elsewhere in this patch, those would become instance properties of a Host object.

I moved the methods into a class that is instantiated by `createHost`. `createHost` further binds the public methods and returns an object with them.

jugglinmike commented:
> While writing this I couldn't respond directly to some of your comments. The reply field is missing for me. But only some of them. So I'll respond here.

Works for me; I'll do the same. I wouldn't be surprised if force-pushing confuses GitHub's comment tracking system. It certainly makes following the development harder for me!

> Changed it to `protocolName`. I tried `name` originally but `name` is already taken. Its the constructor/class name. So I figured should avoid that conflict.

Ah, right, since it's a static property. [I should have recognized that.](https://bocoup.com/blog/whats-in-a-function-name) Anyway, `protocolName` sounds good!

I'm [repeating this next piece of feedback](https://github.com/bocoup/aria-at-automation-harness/pull/2#discussion_r774772377) because, although [you posted a response to it](https://github.com/bocoup/aria-at-automation-harness/pull/2#discussion_r785270234), your explanation was addressing a different suggestion of mine.

> ```js
> function validateUint8Array(bufferData) {
>   invariant(
>     bufferData instanceof Uint8Array,
>     () => `'${bufferData && bufferData.constructor.name}' === 'Uint8Array'`
>   );
>   return bufferData;
> }
> ```

The message-generation logic (rather than the assertion itself) will fail if `bufferData` is a primitive value (or something more exotic like a null-prototyped Object).

> > Until we have real-world experience running the system, I'm considering any protocol-level failure to be an exceptional circumstance--one that maintainers should be informed about immediately.
>
> My main concern for being able to recover is for things arising from new or changed test plans. I expect that the harness will run on aria-at pull requests at some point and so may be fed tests that have some issue, and the automation harness will help identify that. That issue may be a typo from a test author or it could be a driver api and some error on our part we hadn't considered yet. I don't think this at that point yet, but I think that is where its going.

I agree that running tests on pull requests will help identify test bugs, but test semantics and protocol semantics seem to be in completely unrelated domains. For example: I can't think of an ARIA-AT test that would pass when the harness used the "api" protocol and fail when the harness used the "fork" protocol.

When it comes to guarding against errors that we haven't considered yet, that'd be a nice feature, but I don't think we can build the analytics we'd need to deliver on the promise. As you wrote in a subsequent comment, logging will help with forensics--explaining what went wrong when a problem has been identified. But it *won't* help with surfacing issues. Test authors are interested in the accuracy of test results; they have no reason to review debugging messages and determine if they describe a non-critical failure in the infrastructure.

With logging, these errors might not be silent, but they'll sure be quiet. And until someone decides to review debugging data from successful CI tasks (that can be your job :P), we won't even know that part of our system is broken.

> (I think we should make `--debug` default if we detect we are in a CI environment.)

My view is that while such functionality would make the relevant CI script more terse, the system's behavior would be easier to understand if we wrote out `--debug` in that file.

> My editor is vscode with the builtin javascript/typescript functionality. Any tweaks I've made could be replicated by other users by including a `.vscode` directory in the repository that would contain those settings. But that runs up against the dependencies difficulty, which I'm not sure how to resolve without reaching some quorum with more people. (So I'll do just that with an Issue soon.)

I agree that committing this project to a specific editor would be highly irregular and need some pretty convincing justification. What I had in mind was making the TypeScript dependency explicit and including the necessary configuration files for using that tool. Are you open to that?

> > It's now an instance of a Protocol, right? How about `_protocol`?
>
> Erm. Something happened and I somehow ended up editing your comment instead of responding.
>
> What do you think about `_activeProtocol`?

Even better!

jugglinmike reviewed (commented):
[See my prior comment for my review.](https://github.com/bocoup/aria-at-automation-harness/pull/2#issuecomment-1033062721)

mzgoddard commented:
> > ```js
> > function validateUint8Array(bufferData) {
> >     // ...
> >     () => `'${bufferData && bufferData.constructor.name}' === 'Uint8Array'`
> > ```
> 
> The message-generation logic (rather than the assertion itself) will fail if `bufferData` is a primitive value (or something more exotic like a null-prototyped Object).

Fixed!

> With logging, these errors might not be silent, but they'll sure be quiet. And until someone decides to review debugging data from successful CI tasks (that can be your job :P), we won't even know that part of our system is broken.

Removed 'auto' options and related loop logic to try each one. Renamed 'api' choices to 'developer' to try and better illustrate the intended contract.

I'm not tied to the `developer` name for the choice. I thought about calling it `debug` but I'm not sure about that. Like I don't think we want it to be confused as something to use with `--debug`.

-----

Did I get everything?

jugglinmike reviewed (approved):
> Removed 'auto' options and related loop logic to try each one. Renamed 'api' choices to 'developer' to try and better illustrate the intended contract.
> 
> I'm not tied to the `developer` name for the choice. I thought about calling it `debug` but I'm not sure about that. Like I don't think we want it to be confused as something to use with `--debug`.
> 
> Did I get everything?

All except for the TypeScript concern, but like the naming decision, I'm happy to address that in a follow-up. Here are a couple new issues to encourage more input:

- https://github.com/bocoup/aria-at-automation-harness/issues/3
- https://github.com/bocoup/aria-at-automation-harness/issues/4

Nice work, Z!
